### PR TITLE
[SDL-0236] Update mismatch in TireStatus structure

### LINKED ISSUE #339: [SDL 0236] Update mismatch in TireStatus structure
Proposal: [Update mismatch in TireStatus structure](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0231-main-menu-tiles.md)

> This proposal is about correcting a mismatch between HMI_API and MOBILE_API for the `TireStatus` structure.

Review: https://github.com/smartdevicelink/sdl_evolution/issues/743

Steering Committee Decision:

>The Steering Committee voted to accept this proposal with the following revision: clarify that this update will be a breaking change.

> The proposal .md file was updated to reflect these revisions on 6/6/19.

### PR DESCRIPTION
**Pull request by ymalovanyi**
Fixes #339

### Risk
This PR makes **no** API changes.

### Testing Plan
- [x] I have verified that I have not introduced new warnings in this PR (or explain why below)
- [ ] I have run the unit tests with this PR. `N/A`, because no unit tests are needed, as only comments have been changed in the code.
- [ ] I have tested this PR against Core and verified behavior (if applicable, if not applicable, explain why below). `N/A`, because only comments have been changed in the code.

#### Unit Tests
No unit tests are needed, because only comments have been changed in the code.

#### Core Tests
`N/A`

Core version / branch / commit hash / module tested against: `N/A`
HMI name / version / branch / commit hash / module tested against: `N/A`

### Summary
Applied [[SDL-0236] Update mismatch in TireStatus structure](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0236-TireStatus-Mismatch.md) changes

##### Breaking Changes
* `N/A`

##### Enhancements
* Changed comments based on changes in MOBILE_API

##### Bug Fixes
* `N/A`

### Tasks Remaining:
* `N/A`

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

**Comment by ymalovanyi** [comment]
@santhanamk the PR is ready for Ford review.

**Review by santhanamk — COMMENTED** [review]
Hi @ymalovanyi .  I have a question.  I just wanted to check - did you use the code generator tool on the proposal HTML:

`<struct name="TireStatus" since="2.0">
        <description>The status and pressure of the tires.</description>
        <param name="pressureTelltale" type="WarningLightStatus" mandatory="false" since="6.0">
            <description>Status of the Tire Pressure Telltale. See WarningLightStatus.</description>
            <history>
                <param name="pressureTelltale" type="WarningLightStatus" mandatory="true" since="2.0" until="6.0">
            </history>
        </param>
        <param name="leftFront" type="SingleTireStatus" mandatory="false" since="6.0">
            <description>The status of the left front tire.</description>
            <history>
                <param name="leftFront" type="SingleTireStatus" mandatory="true" since="2.0" until="6.0">
            </history>
        </param>
        <param name="rightFront" type="SingleTireStatus" mandatory="false" since="6.0">
            <description>The status of the right front tire.</description>
            <history>
                <param name="rightFront" type="SingleTireStatus" mandatory="true" since="2.0" until="6.0">
            </history>
        </param>
        <param name="leftRear" type="SingleTireStatus" mandatory="false" since="6.0">
            <description>The status of the left rear tire.</description>
            <history>
                <param name="leftRear" type="SingleTireStatus" mandatory="true" since="2.0" until="6.0">
            </history>
        </param>
        <param name="rightRear" type="SingleTireStatus" mandatory="false" since="6.0">
            <description>The status of the right rear tire.</description>
            <history>
                <param name="rightRear" type="SingleTireStatus" mandatory="true" since="2.0" until="6.0">
            </history>
        </param>
        <param name="innerLeftRear" type="SingleTireStatus" mandatory="false" since="6.0">
            <description>The status of the inner left rear.</description>
            <history>
                <param name="innerLeftRear" type="SingleTireStatus" mandatory="true" since="2.0" until="6.0">
            </history>
        </param>
        <param name="innerRightRear" type="SingleTireStatus" mandatory="false" since="6.0">
            <description>The status of the inner right rear.</description>
            <history>
                <param name="innerRightRear" type="SingleTireStatus" mandatory="true" since="2.0" until="6.0">
            </history>
        </param>
    </struct>`

to regenerate TireStatus.js?

**Comment by ymalovanyi** [comment]
Hi @santhanamk 
Yes. The TireStatus.js file is generated with code generator. Only comments have been changed, and changing 'mandatory' attribute doesn't affect any logic in javascript code.

**Review by santhanamk — APPROVED** [review]
Hi @ymalovanyi .  Thanks for the clarification.  I am approving this PR.

**Comment by ymalovanyi** [comment]
@santhanamk, thank you for the approval! @crokita, Ford approved this PR, please review.

**Comment by jordynmackool** [comment]
Hi @ymalovanyi and @santhanamk, since the core PR for this feature is not complete, it seems that this PR would have not been tested fully. To ensure the review process goes smoothly:

1. Please avoid removing sections from the PR template as all the information requested is needed.
2. Please provide the Core branch/PR that this was tested against.

Once this PR is fully tested, tag me and Livio will review. Thank you!

**Comment by vladmu** [comment]
@jordynmackool @crokita the description was fixed to include all sections from the template and the code doesn't require testing against Core nor unit tests (reasons defined in the description). Please review.

**Comment by perryneal11** [comment]
Approved.

**Comment by LitvinenkoIra** [comment]
@theresalech This PR is ready for Livio review. Thank you!

**Comment by crokita** [comment]
@ymalovanyi Hello. There is a breaking change due to the proposal since core no longer needs to send mandatory TireStatus parameters. To avoid causing a breaking change with the next release of the library, TireStatus getters should return default values for current apps not handling null cases. Additionally, a warning should be placed that a future major version change will allow received TireStatus properties to be nullable.

* The pressureTelltale parameter is no longer required and should default to 'NOT_USED'
* Each of the SingleTireStatus parameters are no longer required, but only the inner status property is required. This status value should default to 'UNKNOWN' and the SingleTireStatus object should always exist.

Since the JS suite generates all the RPC structs this change involves modifying the generator's output. My recommendation is to modify the "generator/mapping.json" to remove the default TireStatus getter/setter methods by adding a new object inside the "structs" property:

```js
        "TireStatus": {
            "script": "templates/scripts/TireStatus.js",
            "pressureTelltale": {
                "-methods": {}
            },
            "leftFront": {
                "-methods": {}
            },
            "rightFront": {
                "-methods": {}
            },
            "leftRear": {
                "-methods": {}
            },
            "rightRear": {
                "-methods": {}
            },
            "innerLeftRear": {
                "-methods": {}
            },
            "innerRightRear": {
                "-methods": {}
            }
        }
```

A script can then be created in "generator/templates/scripts/TireStatus.js" that includes those getters and setters back. The setters should be the same, but the getters should enforce non-null returns. Examples below.

```js
/**
 * Get the PressureTelltale
 * @returns {WarningLightStatus} - the KEY_PRESSURE_TELLTALE value
 */
getPressureTelltale () {
    let warningLightStatus = this.getObject(WarningLightStatus, TireStatus.KEY_PRESSURE_TELLTALE);
    if (warningLightStatus === null) {
        warningLightStatus = WarningLightStatus.WLS_NOT_USED;
        this.setParameter(TireStatus.KEY_PRESSURE_TELLTALE, warningLightStatus);
        console.warn('TireStatus\'s pressureTelltale was null and will be set to WarningLightStatus.WLS_NOT_USED. In the future, this will change to be nullable.');
    }
    return warningLightStatus;
}

/**
 * Get the LeftFront
 * @returns {SingleTireStatus} - the KEY_LEFT_FRONT value
 */
getLeftFront () {
    let tireStatus = this.getObject(SingleTireStatus, TireStatus.KEY_LEFT_FRONT);
    if (tireStatus === null) {
        tireStatus = new SingleTireStatus({
            status: 'UNKNOWN',
        });
        this.setParameter(TireStatus.KEY_LEFT_FRONT, tireStatus);
        console.warn('TireStatus\'s leftFront was null and will have its status set to UNKNOWN. In the future, this will change to be nullable.');
    }
    return tireStatus;
}
```

**Comment by YaroslavLutsenko** [comment]
Hello @crokita ! I've processed these changes, please take a look at this (dd79fb2)

File: generator/templates/scripts/TireStatus.js [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+        tireStatus = new SingleTireStatus({
+            status: 'UNKNOWN',
+        });
+        this.setParameter(TireStatus.KEY_INNER_RIGHT_REAR, tireStatus);
+        console.warn('TireStatus\'s innerRightRear was null and will have its status set to UNKNOWN. In the future, this will change to be nullable.');
+    }
+    return tireStatus;
+}
+
+// ----------------- END -----------------------
```
Comment (crokita) [inline generator/templates/scripts/TireStatus.js:195]: Dangling END line comment. This can be removed.

**Review by crokita — CHANGES_REQUESTED** [review]
See minor feedback @YaroslavLutsenko

File: generator/templates/scripts/TireStatus.js (hunk 1 above)
Comment (YaroslavLutsenko) [reply generator/templates/scripts/TireStatus.js]: Removed (d956c3c)

**Comment by YaroslavLutsenko** [comment]
> See minor feedback @YaroslavLutsenko

Thank you for your feedback @crokita.

**Review by crokita — APPROVED** [review]
Looks good.

**Pull request merged**